### PR TITLE
[ATO-716] Add capability to select which custom actions should receive domain

### PR DESCRIPTION
**Pull request by radovanZRasa**
**Proposed changes**:
Add capability to select which custom actions should receive domain.
To enable this capability add _selective_domain_ property under **action_endpoint** in endpoints.yml.
```YAML
# endpoints.yml
action_endpoint:
 url: "http://localhost:5055/webhook"
 enable_selective_domain: true
```
**Once this capability is enabled, domain will be set only to the custom actions which have explicitly stated that they are using domain.**
To specify custom actions which should receive domain go into domain.yml and under actions add {use_domain: True} to those custom actions. For example:
```YAML
# domain.yml
# selective domain is enabled
actions:
  - action_hello_world: {send_domain: True} #will receive domain
  - action_calculate_mass_of_sun # will not receive domain
```
**If selective domain is enabled, any custom action which does not explicitly state that it needs domain, it will not be sent to it.**

To keep this backwards compatible, if selective domain is not enabled then domain will be sent to all custom actions as before.

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by raoulvm** [comment]
I like it. One additional wish though: Can you make it definable with the actions which part of the domain they might receive? 

```yaml
actions:
  - action_hello_world: 
       send_domain: True #will receive full domain
  - action_check_entity:
       send_domain: # will get only parts of the domain
           - entities
           - slots
```

that way you could exclude things you don't want to send.

File: tests/shared/utils/test_validation.py [hunk 1]
```diff
@@ -33,6 +33,17 @@ def test_validate_yaml_schema(file, schema):
     validation_utils.validate_yaml_schema(rasa.shared.utils.io.read_file(file), schema)
 
 
+@pytest.mark.parametrize(
+    "file, schema",
+    [
+        ("data/test_domains/actions.yml", DOMAIN_SCHEMA_FILE),
+    ],
+)
+def test_validate_yaml_schema_actions(file: Text, schema: Text):
```
Comment (sanchariGr) [inline tests/shared/utils/test_validation.py:42]: Looks good, although I remember we had a discussion about possibly adding some kind of integration test for this in the future. If you still think adding one is a good idea, can you please create a ticket in our backlog so we don't forget.

**Comment by sanchariGr** [comment]
Just a comment about the tests and I will leave it to @ancalita to review the implementation:)

File: tests/shared/utils/test_validation.py (hunk 1 above)
Comment (radovanZRasa) [reply tests/shared/utils/test_validation.py:42]: Created: https://rasahq.atlassian.net/browse/ATO-750

**Comment by radovanZRasa** [comment]
> I like it. One additional wish though: Can you make it definable with the actions which part of the domain they might receive?
> 
> ```yaml
> actions:
>   - action_hello_world: 
>        send_domain: True #will receive full domain
>   - action_check_entity:
>        send_domain: # will get only parts of the domain
>            - entities
>            - slots
> ```
> 
> that way you could exclude things you don't want to send.

That is definitely something we have thought of and we will consider as well.
It will take some time to brainstorm how to implement that and in the meanwhile we wanted to be quick and start provide value on this field.

**Review by radovanZRasa — COMMENTED** [review]
@ancalita We should probably put this into 3.4.x first

File: rasa/shared/core/domain.py [hunk 2]
```diff
@@ -1858,6 +1885,34 @@ def __repr__(self) -> Text:
...
+        result: List[Text] = []
+
+        for action in actions:
+            if isinstance(action, dict):
+                for action_name in action:
+                    action_config = action.get(action_name)
+                    if isinstance(action_config, dict):
+                        should_send_domain = action_config.get("send_domain")
+                        if should_send_domain:
+                            result += [action_name]
```
Comment (ancalita) [inline rasa/shared/core/domain.py:1912]: I'd recommend this refactoring as the over-indentation can become difficult to read (plus the schema validation that we have in place would enforce that actions in the domain must be a list of strings or dicts):
```suggestion
                for action_name, action_config in action.items():
                    should_send_domain = action_config.get("send_domain", False)
                    if should_send_domain:
                        result += [action_name]
```

File: rasa/shared/core/domain.py [hunk 3]
```diff
@@ -1858,6 +1885,34 @@ def __repr__(self) -> Text:
...
+                for action_name in action:
+                    result += [action_name]
+            elif isinstance(action, str):
+                result += [action]
+
+        return result
+
+    @staticmethod
+    def _collect_actions_which_explicitly_need_domain(actions: List[Any]) -> List[Text]:
+        result: List[Text] = []
```
Comment (ancalita) [inline rasa/shared/core/domain.py:1903]: We could label these directly `action_names` , it's more descriptive.

File: rasa/shared/core/domain.py [hunk 4]
```diff
@@ -1858,6 +1885,34 @@ def __repr__(self) -> Text:
             f"{len(self.entities)} entities, {len(self.form_names)} forms"
         )
 
+    @staticmethod
+    def _collect_actions(actions: List[Any]) -> List[Text]:
+        result: List[Text] = []
```
Comment (ancalita) [inline rasa/shared/core/domain.py:1890]: We could label this variable also with `action_names`.

File: rasa/shared/core/domain.py [hunk 5]
```diff
@@ -1849,6 +1861,21 @@ def count_slot_mapping_statistics(self) -> Tuple[int, int, int]:
 
         return (total_mappings, custom_mappings, conditional_mappings)
 
+    def does_action_explicitly_need_domain(self, action_name: Text) -> bool:
```
Comment (ancalita) [inline rasa/shared/core/domain.py:1864]:
```suggestion
    def does_custom_action_explicitly_need_domain(self, action_name: Text) -> bool:
```

File: rasa/shared/core/domain.py [hunk 6]
```diff
@@ -1849,6 +1861,21 @@ def count_slot_mapping_statistics(self) -> Tuple[int, int, int]:
...
+
+        Args:
+            action_name: Name of the action to be checked
+
+        Returns:
+            True if action has explicitly stated that it needs domain.
+            Otherwise, it returns false.
+        """
+        return (
+            self._actions_which_explicitly_need_domain is not None
```
Comment (ancalita) [inline rasa/shared/core/domain.py:1875]: this attribute is never None though, it's either an empty list or a list with strings.

File: rasa/shared/core/domain.py [hunk 7]
```diff
@@ -239,6 +239,11 @@ def from_dict(cls, data: Dict) -> "Domain":
         if domain_slots:
             rasa.shared.core.slot_mappings.validate_slot_mappings(domain_slots)
         slots = cls.collect_slots(domain_slots)
+        domain_actions = data.get(KEY_ACTIONS, [])
+        actions = cls._collect_actions(domain_actions)
+        actions_which_explicitly_need_domain = (
+            cls._collect_actions_which_explicitly_need_domain(domain_actions)
+        )
```
Comment (ancalita) [inline rasa/shared/core/domain.py:246]: I'd rather recommend moving this logic into the constructor, rather than modifying the signature of the constructor (it's already super long 😅 )

File: rasa/core/actions/action.py [hunk 8]
```diff
@@ -636,27 +637,48 @@ def __init__(self, name: Text, action_endpoint: Optional[EndpointConfig]) -> Non
...
+    def _is_selective_domain_enabled(self) -> bool:
+        is_selective_domain_enabled = False
+        if self.action_endpoint is not None:
+            is_selective_domain_enabled = bool(
+                self.action_endpoint.kwargs.get(
+                    SELECTIVE_DOMAIN, DEFAULT_SELECTIVE_DOMAIN
+                )
+            )
+
+        return is_selective_domain_enabled
```
Comment (ancalita) [inline rasa/core/actions/action.py:673]: I'd refactor this too, to simplify overindentation (probably needs some extra formatting again):
```suggestion
        if self.action_endpoint is None:
            # add a logger.debug to warn of this issue
            return False
        return bool(
                self.action_endpoint.kwargs.get(
                    SELECTIVE_DOMAIN, DEFAULT_SELECTIVE_DOMAIN
                )
            )
```

File: tests/shared/core/test_domain.py [hunk 9]
```diff
@@ -2067,3 +2067,112 @@ def test_merge_domain_with_separate_session_config():
...
+                {"action_hello_world": {"send_domain": False}},
+                {"action_say_something": {"send_domain": True}},
+                {"action_calculate": {"send_domain": True}},
+                "action_no_domain",
+            ],
+            ["action_say_something", "action_calculate"],
+        )
+    ],
+)
+def test_collect_actions_which_explicitly_need_domain(
```
Comment (ancalita) [inline tests/shared/core/test_domain.py:2086]: Could we add more edge cases here (when actions list is empty, when actions list contains only strings etc)

File: tests/shared/core/test_domain.py [hunk 10]
```diff
@@ -2067,3 +2067,112 @@ def test_merge_domain_with_separate_session_config():
...
+    # assert that two unordered lists have same elements
+    assert sorted(result) == sorted(expected_result)
+
+
+@pytest.mark.parametrize(
+    "content, expected_user_actions, expected_actions_which_explicitly_need_domain",
+    [
+        (
+            """
+        intents:
```
Comment (ancalita) [inline tests/shared/core/test_domain.py:2128]: Please add `version` key to the domain contents to prevent overflow of UserWarnings in the CI.

**Review by ancalita — COMMENTED** [review]
Looks great 💯 Left some suggestions here and there.

File: tests/shared/core/test_domain.py [hunk 11]
```diff
@@ -2067,3 +2067,134 @@ def test_merge_domain_with_separate_session_config():
...
+          - action_hello: {{send_domain: True}}
+          - action_bye: {{send_domain: True}}
+          - action_no_domain
+          """,
+            ["action_hello", "action_bye", "action_no_domain"],
+            ["action_hello", "action_bye"],
+        ),
+        (
+            """
+        intents:
```
Comment (ancalita) [inline tests/shared/core/test_domain.py:2170]: Version key is missing here too 😄

**Review by ancalita — APPROVED** [review]
Almost there 🎉 Please make the docs-related changes 🙏🏻

File: rasa/core/actions/action.py [hunk 12]
```diff
@@ -636,27 +637,45 @@ def __init__(self, name: Text, action_endpoint: Optional[EndpointConfig]) -> Non
...
+            not self._is_selective_domain_enabled()
+            or domain.does_custom_action_explicitly_need_domain(self.name())
+        ):
+            result["domain"] = domain.as_dict()
+
+        return result
+
+    def _is_selective_domain_enabled(self) -> bool:
+        if self.action_endpoint is None:
+            # add a logger.debug to warn of this issue
```
[the remark this answers is not in the document]
Comment (ancalita) [reply rasa/core/actions/action.py:666]: Or maybe `raise_warning` is more appropriate as it seems more of a usage error.

[the remark this answers is not in the document]
Comment (radovanZRasa) [reply rasa/core/actions/action.py:666]: This comment is just junk.

**Comment by radovanZRasa** [comment]
I closed this PR as we have merged 3.4.x into main and it already contains this same change through this PR https://github.com/RasaHQ/rasa/pull/12026
